src/lib.rs

```diff
@@ -578,32 +643,93 @@ pub fn max_log_level() -> LogLevelFilter {
-    let logger = unsafe { mem::transmute::<Box<Box<Log>>, usize>(logger) };
-    LOGGER.store(logger, Ordering::SeqCst);
+    extern fn shutdown() {
+        shutdown_logger().map(|logger| unsafe {
+            Box::from_raw(logger as *mut Log);
+        }).ok();
+    }
+}
 
-    unsafe {
-        assert_eq!(libc::atexit(shutdown), 0);
+/// Sets the global logger from a raw pointer.
+///
+/// This function is similar to `set_logger` except that it is usable in
+/// `no_std` code. Another difference is that the logger is not automatically
+/// shut down on program exit, and `shutdown_logger` must be called to manually
+/// shut it down.
+///
+/// The `make_logger` closure is passed a `MaxLogLevel` object, which the
+/// logger should use to keep the global maximum log level in sync with the
+/// highest log level that the logger will not ignore.
+///
+/// This function may only be called once in the lifetime of a program. Any log
+/// events that occur before the call to `set_logger_raw` completes will be
+/// ignored.
+///
+/// This function does not typically need to be called manually. Logger
+/// implementations should provide an initialization method that calls
+/// `set_logger_raw` internally.
+///
+/// # Safety
+///
+/// The pointer returned by `make_logger` must remain valid for the entire
+/// duration of the program or until `shutdown_logger` is called.
+pub unsafe fn set_logger_raw<M>(make_logger: M) -> Result<(), SetLoggerError>
+        where M: FnOnce(MaxLogLevelFilter) -> *const Log {
+    if STATE.compare_and_swap(UNINITIALIZED, INITIALIZING,
+                              Ordering::SeqCst) != UNINITIALIZED {
+        return Err(SetLoggerError(()));
     }
-    return Ok(());
 
-    extern fn shutdown() {
-        // Set to INITIALIZING to prevent re-initialization after
-        let logger = LOGGER.swap(INITIALIZING, Ordering::SeqCst);
+    LOGGER = Some(make_logger(MaxLogLevelFilter(())));
+    STATE.store(INITIALIZED, Ordering::SeqCst);
+    Ok(())
+}
 
-        while REFCOUNT.load(Ordering::SeqCst) != 0 {
-            // FIXME add a sleep here when it doesn't involve timers
-        }
+/// Shuts down the global logger.
+///
+/// This function may only be called once in the lifetime of a program, and may
+/// not be called before `set_logger_raw`. Once the global logger has been shut
+/// down, it can no longer be re-initialized by `set_logger_raw`. Any log
+/// events that occur after the call to `shutdown_logger` completes will be
+/// ignored.
+///
+/// The pointer that was originally passed to `set_logger_raw` is returned on
+/// success. At that point it is guaranteed that no other threads are
+/// concurrently accessing the logger object.
+///
+/// This function should not be called when the global logger was registered
+/// using `set_logger`, since in that case the logger will automatically be shut
+/// down when the program exits
+pub fn shutdown_logger() -> Result<*const Log, ShutdownLoggerError> {
```

Comment:
I think a module perhaps, but it seems kinda arbitrary. Including raw in the name seems prudent though.